File: backend/app/api/analytics.py
from datetime import date
from decimal import Decimal

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.schemas import (
    AnalyticsBreakdownItem,
    AnalyticsDistributionResponse,
    AnalyticsFilteredResponse,
    AnalyticsResponse,
    AnalyticsSummary,
    AnalyticsTrendResponse,
)
from app.services.analytics_service import AnalyticsService

router = APIRouter(prefix="/analytics", tags=["analytics"])


def _date_filters(
    as_of_date: date | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
) -> date:
    """Resolve snapshot date: as_of_date, or date_to from range, or today."""
    if as_of_date:
        return as_of_date
    if date_to:
        return date_to
    return date.today()


def _scope_filters(
    country_ids: list[int] | None = None,
    department_ids: list[int] | None = None,
    job_level_ids: list[int] | None = None,
    payment_statuses: list[str] | None = None,
    employment_statuses: list[str] | None = None,
    record_statuses: list[str] | None = None,
    currency_ids: list[int] | None = None,
    salary_min: float | None = None,
    salary_max: float | None = None,
) -> dict:
    return {
        "country_ids": country_ids or None,
        "department_ids": department_ids or None,
        "job_level_ids": job_level_ids or None,
        "payment_statuses": payment_statuses or None,
        "employment_statuses": employment_statuses or None,
        "record_statuses": record_statuses or None,
        "currency_ids": currency_ids or None,
        "salary_min": Decimal(str(salary_min)) if salary_min is not None else None,
        "salary_max": Decimal(str(salary_max)) if salary_max is not None else None,
    }


def _range_or_snapshot_kwargs(
    as_of_date: date | None,
    date_from: date | None,
    date_to: date | None,
) -> dict:
    if date_from and date_to:
        return {"date_from": date_from, "date_to": date_to}
    return {"as_of_date": _date_filters(as_of_date, date_from, date_to)}


@router.get("/filtered", response_model=AnalyticsFilteredResponse)
def get_analytics_filtered(
    reporting_currency: str = Query(default="USD"),
    as_of_date: date | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    country_ids: list[int] | None = Query(None),
    department_ids: list[int] | None = Query(None),
    job_level_ids: list[int] | None = Query(None),
    payment_statuses: list[str] | None = Query(None),
    employment_statuses: list[str] | None = Query(None),
    record_statuses: list[str] | None = Query(None),
    currency_ids: list[int] | None = Query(None),
    salary_min: float | None = None,
    salary_max: float | None = None,
    db: Session = Depends(get_db),
):
    """All analytics for the exact filter combination in one response."""
    service = AnalyticsService(db)
    scope = _scope_filters(
        country_ids,
        department_ids,
        job_level_ids,
        payment_statuses,
        employment_statuses,
        record_statuses,
        currency_ids,
        salary_min,
        salary_max,
    )
    if date_from and date_to:
        return service.get_filtered(
            reporting_currency,
            date_from=date_from,
            date_to=date_to,
            **scope,
        )
    return service.get_filtered(
        reporting_currency,
        as_of_date=_date_filters(as_of_date, date_from, date_to),
        **scope,
    )


@router.get("/summary", response_model=AnalyticsSummary)
def get_analytics_summary(
    reporting_currency: str = Query(default="USD"),
    as_of_date: date | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    country_ids: list[int] | None = Query(None),
    department_ids: list[int] | None = Query(None),
    job_level_ids: list[int] | None = Query(None),
    payment_statuses: list[str] | None = Query(None),
    employment_statuses: list[str] | None = Query(None),
    record_statuses: list[str] | None = Query(None),
    currency_ids: list[int] | None = Query(None),
    salary_min: float | None = None,
    salary_max: float | None = None,
    db: Session = Depends(get_db),
):
    service = AnalyticsService(db)
    scope = _scope_filters(
        country_ids,
        department_ids,
        job_level_ids,
        payment_statuses,
        employment_statuses,
        record_statuses,
        currency_ids,
        salary_min,
        salary_max,
    )
    if date_from and date_to:
        return service.get_summary(
            reporting_currency,
            date_from=date_from,
            date_to=date_to,
            **scope,
        )
    snapshot = _date_filters(as_of_date, date_from, date_to)
    return service.get_summary(
        reporting_currency,
        as_of_date=snapshot,
        **scope,
    )


@router.get("/by-country", response_model=list[AnalyticsBreakdownItem])
def get_analytics_by_country(
    reporting_currency: str = Query(default="USD"),
    as_of_date: date | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    country_ids: list[int] | None = Query(None),
    department_ids: list[int] | None = Query(None),
    job_level_ids: list[int] | None = Query(None),
    payment_statuses: list[str] | None = Query(None),
    employment_statuses: list[str] | None = Query(None),
    record_statuses: list[str] | None = Query(None),
    currency_ids: list[int] | None = Query(None),
    salary_min: float | None = None,
    salary_max: float | None = None,
    db: Session = Depends(get_db),
):
    service = AnalyticsService(db)
    return service.get_by_country(
        reporting_currency,
        **_range_or_snapshot_kwargs(as_of_date, date_from, date_to),
        **_scope_filters(
            country_ids,
            department_ids,
            job_level_ids,
            payment_statuses,
            employment_statuses,
            record_statuses,
            currency_ids,
            salary_min,
            salary_max,
        ),
    )


@router.get("/by-department", response_model=list[AnalyticsBreakdownItem])
def get_analytics_by_department(
    reporting_currency: str = Query(default="USD"),
    as_of_date: date | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    country_ids: list[int] | None = Query(None),
    department_ids: list[int] | None = Query(None),
    job_level_ids: list[int] | None = Query(None),
    payment_statuses: list[str] | None = Query(None),
    employment_statuses: list[str] | None = Query(None),
    record_statuses: list[str] | None = Query(None),
    currency_ids: list[int] | None = Query(None),
    salary_min: float | None = None,
    salary_max: float | None = None,
    db: Session = Depends(get_db),
):
    service = AnalyticsService(db)
    return service.get_by_department(
        reporting_currency,
        **_range_or_snapshot_kwargs(as_of_date, date_from, date_to),
        **_scope_filters(
            country_ids,
            department_ids,
            job_level_ids,
            payment_statuses,
            employment_statuses,
            record_statuses,
            currency_ids,
            salary_min,
            salary_max,
        ),
    )


@router.get("/by-level", response_model=list[AnalyticsBreakdownItem])
def get_analytics_by_level(
    reporting_currency: str = Query(default="USD"),
    as_of_date: date | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    country_ids: list[int] | None = Query(None),
    department_ids: list[int] | None = Query(None),
    job_level_ids: list[int] | None = Query(None),
    payment_statuses: list[str] | None = Query(None),
    employment_statuses: list[str] | None = Query(None),
    record_statuses: list[str] | None = Query(None),
    currency_ids: list[int] | None = Query(None),
    salary_min: float | None = None,
    salary_max: float | None = None,
    db: Session = Depends(get_db),
):
    service = AnalyticsService(db)
    return service.get_by_level(
        reporting_currency,
        **_range_or_snapshot_kwargs(as_of_date, date_from, date_to),
        **_scope_filters(
            country_ids,
            department_ids,
            job_level_ids,
            payment_statuses,
            employment_statuses,
            record_statuses,
            currency_ids,
            salary_min,
            salary_max,
        ),
    )


@router.get("/distribution", response_model=AnalyticsDistributionResponse)
def get_analytics_distribution(
    reporting_currency: str = Query(default="USD"),
    as_of_date: date | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    country_ids: list[int] | None = Query(None),
    department_ids: list[int] | None = Query(None),
    job_level_ids: list[int] | None = Query(None),
    payment_statuses: list[str] | None = Query(None),
    employment_statuses: list[str] | None = Query(None),
    record_statuses: list[str] | None = Query(None),
    currency_ids: list[int] | None = Query(None),
    salary_min: float | None = None,
    salary_max: float | None = None,
    db: Session = Depends(get_db),
):
    service = AnalyticsService(db)
    return service.get_distribution(
        reporting_currency,
        **_range_or_snapshot_kwargs(as_of_date, date_from, date_to),
        **_scope_filters(
            country_ids,
            department_ids,
            job_level_ids,
            payment_statuses,
            employment_statuses,
            record_statuses,
            currency_ids,
            salary_min,
            salary_max,
        ),
    )


@router.get("/trend", response_model=AnalyticsTrendResponse)
def get_analytics_trend(
    reporting_currency: str = Query(default="USD"),
    date_from: date = Query(...),
    date_to: date = Query(...),
    country_ids: list[int] | None = Query(None),
    department_ids: list[int] | None = Query(None),
    job_level_ids: list[int] | None = Query(None),
    payment_statuses: list[str] | None = Query(None),
    employment_statuses: list[str] | None = Query(None),
    record_statuses: list[str] | None = Query(None),
    currency_ids: list[int] | None = Query(None),
    salary_min: float | None = None,
    salary_max: float | None = None,
    db: Session = Depends(get_db),
):
    service = AnalyticsService(db)
    return service.get_trend(
        reporting_currency,
        date_from,
        date_to,
        **_scope_filters(
            country_ids,
            department_ids,
            job_level_ids,
            payment_statuses,
            employment_statuses,
            record_statuses,
            currency_ids,
            salary_min,
            salary_max,
        ),
    )


@router.get("", response_model=AnalyticsResponse, include_in_schema=False)
def get_analytics_legacy(
    reporting_currency: str = Query(default="USD"),
    as_of_date: date | None = None,
    country_ids: list[int] | None = Query(None),
    department_ids: list[int] | None = Query(None),
    job_level_ids: list[int] | None = Query(None),
    payment_statuses: list[str] | None = Query(None),
    employment_statuses: list[str] | None = Query(None),
    record_statuses: list[str] | None = Query(None),
    currency_ids: list[int] | None = Query(None),
    db: Session = Depends(get_db),
):
    service = AnalyticsService(db)
    return service.get_analytics(
        reporting_currency,
        as_of_date,
        **_scope_filters(
            country_ids,
            department_ids,
            job_level_ids,
            payment_statuses,
            employment_statuses,
            record_statuses,
            currency_ids,
        ),
    )

File: frontend/src/shared/api/analytics.ts
import { buildQuery, fetchApi } from './http';

export interface AnalyticsSummary {
  employee_count: number;
  average_compensation: number;
  median_compensation: number;
  min_compensation: number;
  max_compensation: number;
  total_compensation: number;
  reporting_currency: string;
  as_of_date?: string | null;
}

export interface AnalyticsTrendPoint {
  period_label: string;
  period_date: string;
  employee_count: number;
  average_compensation: number;
  total_compensation: number;
}

export interface AnalyticsTrend {
  reporting_currency: string;
  date_from: string;
  date_to: string;
  points: AnalyticsTrendPoint[];
}

export interface AnalyticsFilters {
  as_of_date?: string;
  date_from?: string;
  date_to?: string;
  department_ids?: number[];
  job_level_ids?: number[];
  payment_statuses?: string[];
  employment_statuses?: string[];
  record_statuses?: string[];
  currency_ids?: number[];
  salary_min?: number;
  salary_max?: number;
}

export interface AnalyticsPaymentSummary {
  total_records: number;
  employee_count: number;
  paid_count: number;
  not_paid_count: number;
  total_paid_amount: number;
  total_not_paid_amount: number;
  reporting_currency: string;
}

export interface AnalyticsBreakdown {
  group_key: string;
  group_label: string;
  employee_count: number;
  average_compensation: number;
  total_compensation: number;
}

export interface DistributionBucket {
  range_label: string;
  min_value: number;
  max_value: number | null;
  employee_count: number;
  percentage: number;
}

export interface AnalyticsDistribution {
  reporting_currency: string;
  total_employees: number;
  buckets: DistributionBucket[];
}

export interface Analytics {
  summary: AnalyticsSummary;
  by_country: AnalyticsBreakdown[];
  by_department: AnalyticsBreakdown[];
  by_job_level: AnalyticsBreakdown[];
}

export interface AnalyticsFiltered {
  reporting_currency: string;
  mode: 'snapshot' | 'range';
  filters: AnalyticsFilters;
  summary: AnalyticsSummary;
  payment_summary: AnalyticsPaymentSummary;
  range_start_summary: AnalyticsSummary | null;
  range_end_summary: AnalyticsSummary | null;
  by_country: AnalyticsBreakdown[];
  by_department: AnalyticsBreakdown[];
  by_job_level: AnalyticsBreakdown[];
  distribution: AnalyticsDistribution;
  trend: AnalyticsTrend | null;
}

export const analyticsApi = {
  getAnalyticsSummary: (reportingCurrency: string, filters?: AnalyticsFilters) =>
    fetchApi<AnalyticsSummary>(`/analytics/summary${buildQuery({ reporting_currency: reportingCurrency, ...filters })}`),

  getAnalyticsByCountry: (reportingCurrency: string, filters?: AnalyticsFilters) =>
    fetchApi<AnalyticsBreakdown[]>(`/analytics/by-country${buildQuery({ reporting_currency: reportingCurrency, ...filters })}`),

  getAnalyticsByDepartment: (reportingCurrency: string, filters?: AnalyticsFilters) =>
    fetchApi<AnalyticsBreakdown[]>(`/analytics/by-department${buildQuery({ reporting_currency: reportingCurrency, ...filters })}`),

  getAnalyticsByLevel: (reportingCurrency: string, filters?: AnalyticsFilters) =>
    fetchApi<AnalyticsBreakdown[]>(`/analytics/by-level${buildQuery({ reporting_currency: reportingCurrency, ...filters })}`),

  getAnalyticsDistribution: (reportingCurrency: string, filters?: AnalyticsFilters) =>
    fetchApi<AnalyticsDistribution>(`/analytics/distribution${buildQuery({ reporting_currency: reportingCurrency, ...filters })}`),

  getAnalyticsTrend: (
    reportingCurrency: string,
    dateFrom: string,
    dateTo: string,
    filters?: Omit<AnalyticsFilters, 'date_from' | 'date_to' | 'as_of_date'>,
  ) =>
    fetchApi<AnalyticsTrend>(`/analytics/trend${buildQuery({ reporting_currency: reportingCurrency, date_from: dateFrom, date_to: dateTo, ...filters })}`),

  getFilteredAnalytics: (reportingCurrency: string, filters?: AnalyticsFilters) =>
    fetchApi<AnalyticsFiltered>(`/analytics/filtered${buildQuery({ reporting_currency: reportingCurrency, ...filters })}`),

  getAnalytics: async (reportingCurrency: string, filters?: AnalyticsFilters): Promise<Analytics> => {
    const query = buildQuery({ reporting_currency: reportingCurrency, ...filters });
    const [summary, by_country, by_department, by_job_level] = await Promise.all([
      fetchApi<AnalyticsSummary>(`/analytics/summary${query}`),
      fetchApi<AnalyticsBreakdown[]>(`/analytics/by-country${query}`),
      fetchApi<AnalyticsBreakdown[]>(`/analytics/by-department${query}`),
      fetchApi<AnalyticsBreakdown[]>(`/analytics/by-level${query}`),
    ]);
    return { summary, by_country, by_department, by_job_level };
  },
};
